### ISSUE TITLE
Support numBits on homogeneous tuples 

### ISSUE THREAD

**Issue by ben-albrecht** [chapel-lang/chapel#6915]
### Idea

It would be nice to support `numBits()` on homogeneous tuples, such that the following program would work:

```chpl
writeln(numBits(2*int));
```

### Motivation

I encountered this shortcoming when trying to create a `RandomStream(eltType=2*int)` for generating random 2D array indices.

### Implementation

This should be possible to implement with the following additions to `ChapelTuple.chpl`:

- `_tuple.size param` (on the *type*)
- `isHomogeneousTupleType(type t) param`
- `homogeneousTupleType(type t) param`

Due to the special-casing of tuples, these would require some compiler modifications.

The overload might then look like:

```chpl
pragma "no doc"
proc numBits(type t) param where isHomogeneousType(t) {
 return t.size*homogeneousTupleType(t);
}
```

**Comment by bradcray** [comment]
It sounds to me like your issue is more about wanting `RandomStream` to support tuples of ints than about `numBits()` supporting tuples.  It's not obvious to me that it's appropriate for `RandomStream` to use `numBits()` in this case even if it were supported on tuples.  If you replace that expression with the integer you expect, does everything else compile and run as expected? (my bet would be "no", but I didn't try it).  In either case, a better approach might be to have that routine handle the tuple case explicitly.

Historically, I believe that `numBits()` has been about the number of bits that a type requires in physical memory at execution time which is why we don't typically support it for records and classes (say) where it's hard to predict at compile-time.  Thinking about our back-end implementations (static arrays in C; I'm not sure in LLVM), I can't recall C's memory layout rules well enough to remember whether we can guarantee that the elements are densely packed in memory (as your implementation suggests) or not.  This might be a question for @dmk42.

**Comment by dmk42** [comment]
Funny that this came up.  I just ran into the size issue while working on something different.  For the reasons Brad explains, the Chapel compiler does not know the size of a class when the back end is C.  When the back end is LLVM, there is a little more information available because the type can be converted to an LLVM type and then the size read back from there.

**Comment by bradcray** [comment]
But @dmk42:  Can a static C array's size be predicted if one knows the element type?  I.e., are static C arrays guaranteed to be consecutive and packed in memory (in which case numElts * eltSize would work), or could the C compiler do something different? (in which case we couldn't predict, as with structs).  Thanks.

**Comment by dmk42** [comment]
Arrays are consecutively packed.  Padding is added to the end of structs to ensure that an array of structs can be consecutively packed without violating alignment constraints.

**Comment by bradcray** [comment]
Great, thanks.  In that case, it seems as though we could support `numBits()` on _homogeneous_ tuple types whose element type also supported `numBits()`.  Unless the LLVM back-end did something different here...

(All that said, I still strongly suspect that the RandomStream module doesn't actually want to be reasoning about the memory footprint of these types...).

**Comment by dmk42** [comment]
The LLVM back end intends to get the same result.  If we find a case where it doesn't, it's a bug.

**Comment by ben-albrecht** [comment]
> It sounds to me like your issue is more about wanting RandomStream to support tuples of ints than about numBits() supporting tuples.

While `RandomStream` is the motivating case that led me to this issue, I still think `numBits()` supporting *homogenous* tuples would be useful.

Perhaps `RandomStream` support for tuples should be a separate issue.

**Comment by bradcray** [comment]
> While RandomStream is the motivating case that led me to this issue, I still think numBits() supporting homogenous tuples would be useful.

Yep, and based on the discussion above, I think you could go ahead and add it with little difficulty.

**Comment by ben-albrecht** [comment]
>  I think you could go ahead and add it with little difficulty.

Sweet - thanks!
